sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/FailFastRntbdRequestRecord.java

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.cosmos.implementation.directconnectivity.rntbd;
+
+import com.azure.cosmos.BridgeInternal;
+import com.azure.cosmos.implementation.HttpConstants;
+import com.azure.cosmos.implementation.RequestTimeoutException;
+import com.azure.cosmos.implementation.http.HttpHeaders;
+import io.netty.util.Timeout;
+import io.netty.util.TimerTask;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.SocketAddress;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static com.azure.cosmos.implementation.guava27.Strings.lenientFormat;
+
+public class FailFastRntbdRequestRecord extends RntbdRequestRecord {
+    private static final Logger logger = LoggerFactory.getLogger(RntbdRequestRecord.class);
+
+    private FailFastRntbdRequestRecord(final RntbdRequestArgs args) {
+        super(args);
+    }
+
+    @Override
+    public Timeout newTimeout(final TimerTask task) {
+        throw new IllegalArgumentException("newTimeout must never be called for fail fast records.");
+    }
+
+    public static FailFastRntbdRequestRecord createAndFailFast(
+        RntbdRequestArgs args,
+        long concurrentRequestsSnapshot,
+        RntbdMetrics metrics,
+        SocketAddress remoteAddress) {
+
+        FailFastRntbdRequestRecord failFastRecord = new FailFastRntbdRequestRecord(args);
+
+        logger.debug(
```

Comment:
The usual pattern I saw is: (should we follow here as well?)
 if (logger.isDebugEnabled) {
     logger.debug()
}

Reply:
The pattern is used extremely inconsistent - from my understanding it is extremely useful if it is expensive to "calculate" the message to be logged (to avoid even preparing the message when debug logging is disabled). I will leave it here as is (cheap to get the to be logged message). But I agree that it would be good to go over the entire code base and decide how to get this consistent. I can pick this up when I look at the perf optimizations that I want to take a look at (primarily for write)

Reply:
For admission control with lots of load its useful.